contracts/FigmentEth2Depositor.t.sol

```diff
@@ -2,14 +2,954 @@
+            _createValidValidatorData(32_000_000_000);
+        amountsGwei[0] = 32_000_000_000;
+
+        vm.expectRevert(
+            abi.encodeWithSelector(FigmentEth2Depositor.InvalidValidatorData.selector, 0, "withdrawal_credentials")
+        );
+        figmentDepositor.deposit{value: 32 ether}(
+            pubkeys, wrongWithdrawalCredentials, signatures, depositDataRoots, amountsGwei
+        );
+    }
+
+    function test_Deposit_InvalidSignatureLength_Reverts() public {
+        bytes[] memory pubkeys = new bytes[](1);
+        bytes[] memory withdrawalCredentials = new bytes[](1);
+        bytes[] memory signatures = new bytes[](1);
+        bytes32[] memory depositDataRoots = new bytes32[](1);
+        uint256[] memory amountsGwei = new uint256[](1);
+
+        // Create signature with wrong length (95 bytes instead of 96)
+        bytes[] memory wrongSignatures = new bytes[](1);
+        wrongSignatures[0] = new bytes(95);
+        for (uint256 i = 0; i < 95; i++) {
+            wrongSignatures[0][i] = bytes1(uint8(i % 256));
+        }
+
+        (pubkeys[0], withdrawalCredentials[0], signatures[0], depositDataRoots[0]) =
+            _createValidValidatorData(32_000_000_000);
+        amountsGwei[0] = 32_000_000_000;
+
+        vm.expectRevert(abi.encodeWithSelector(FigmentEth2Depositor.InvalidValidatorData.selector, 0, "signature"));
+        figmentDepositor.deposit{value: 32 ether}(
+            pubkeys, withdrawalCredentials, wrongSignatures, depositDataRoots, amountsGwei
+        );
+    }
+
+    function test_Deposit_MultipleValidatorsDataValidation() public {
+        bytes[] memory pubkeys = new bytes[](2);
+        bytes[] memory withdrawalCredentials = new bytes[](2);
+        bytes[] memory signatures = new bytes[](2);
+        bytes32[] memory depositDataRoots = new bytes32[](2);
+        uint256[] memory amountsGwei = new uint256[](2);
+
+        // First validator with valid data
+        (pubkeys[0], withdrawalCredentials[0], signatures[0], depositDataRoots[0]) =
+            _createValidValidatorData(32_000_000_000);
+        amountsGwei[0] = 32_000_000_000;
+
+        // Second validator with invalid pubkey length
+        pubkeys[1] = new bytes(47); // Wrong length
+        for (uint256 i = 0; i < 47; i++) {
+            pubkeys[1][i] = bytes1(uint8(i % 256));
+        }
+        (withdrawalCredentials[1], signatures[1], depositDataRoots[1]) =
+            _createValidValidatorDataWithoutPubkey(32_000_000_000);
+        amountsGwei[1] = 32_000_000_000;
+
+        uint256 totalValue = 64 ether;
+
+        vm.expectRevert(abi.encodeWithSelector(FigmentEth2Depositor.InvalidValidatorData.selector, 1, "pubkey"));
+        figmentDepositor.deposit{value: totalValue}(
+            pubkeys, withdrawalCredentials, signatures, depositDataRoots, amountsGwei
+        );
+    }
+
+    // ============ Deposits are all or nothing Tests ============
+
```

Comment:
The test below is new and confirms that if one deposit fails everything reverts.